### ISSUE TITLE
Deprecating WebKit1 label:"help wanted"

### ISSUE THREAD

**Issue by pydoh** [r0x0r/pywebview#210]
### Specification
- Platform: Fedora
- Version: 27

### Description
I was wondering if you are going to move to WebKit2 and deprecate WebKit1. I've been trying to develop a Python/Gtk application and keep seeing warnings about WebKit security.
Such as;
https://blogs.gnome.org/mcatanzaro/2016/02/01/on-webkit-security-updates/

In my application I'm using a try/except to test the version like this;
`
import gi
try:
    gi.require_version('Gtk', '3.0')
    from gi.repository import Gtk, Gdk  # , Gio
except:
    sys.exit("This script requires Gtk 3.0 or newer!")

try:
    gi.require_version('WebKit2', '4.0')
    from gi.repository import WebKit2
except:
    sys.exit("This script requires WebKit2 4.0 or newer!")
`
As a side note, I'm having a difficult time learning how to use WebKit. Searching for answers is what led me to pywebview.

**Comment by shivaprsdv** [comment]
> I was wondering if you are going to move to WebKit2 and deprecate
> WebKit1.

There has been no discussion on this, so far I can see. While it is a good
suggestion, it seems like this would need a lot of changes at the core, it
could need some work. So I don't see this could be done any soon.

Likewise, the WebView on macOS has also become legacy, replaced by
WKWebView, which Apple encourages to be used for macOS 10.10+.

> I'm having a difficult time learning how to use WebKit. Searching for
> answers is what led me to pywebview.

Try also reading the original [C documentation](https://webkitgtk.org/reference/webkitgtk/stable/webkitgtk-webkitwebview.html). Translating it to Python is
not hard once you get it. The above doc is for WebKit1, but the basic
architecture ought to be similar on WebKit2 as well.

> Also, I tried to find a way to label this as label:"help wanted", but I
> can't see how to apply a label, sorry.

Only the project owner and maintainers can apply labels. Anyway the
`help wanted` label is applied when _we_ want help from others on an
issue. _You_ don't need to put the label for us to help you. 🙂

**Comment by r0x0r** [comment]
To be honest, this is the first time I hear about WebKit2. I am all up for an upgrade.

**Comment by shivaprsdv** [comment]
@r0x0r [WebKit2](https://en.wikipedia.org/wiki/WebKit#WebKit2) [WebKit2GTK+](https://webkitgtk.org/reference/webkit2gtk/stable/index.html) [WebKitWebView](https://webkitgtk.org/reference/webkit2gtk/stable/WebKitWebView.html).

Well, then we should also consider upgrading to WKWebView on Cocoa. Looks
like a job after v2.0 release, though.

**Comment by r0x0r** [comment]
I gave WebKit2 a quick test and it seems to work more or less out of the box apart from the fact they have changed the names of the properties and functions of webview. E.g. `execute_script` is now `run_javascript` (it returns a value as well!). Anyhow it should not be a big job  to perform this upgrade.  GTK documentation is horrendous, though

**Comment by shivaprsdv** [comment]
Right. So is it before or after Pywebview 2? I guess I can dig into macOS.

**Comment by r0x0r** [comment]
After.

**Comment by francescomeloni** [comment]
hi, any news about that?

**Comment by r0x0r** [comment]
I don't really do any development in the summer, so I will probably be
able to give it a go sometime in the fall.
On that note, be the change you want to see in the world.

**Comment by r0x0r** [comment]
I have started GTK upgrade in the `webkit2` branch. The first hurdle is that JS result cannot be returned from `run_javascript` as the GTK Python port does not support it. A C library exists for this purpose, but I would not like to introduce it as a dependency.

Looks like manipulating the window title is still the way to go.

**Comment by r0x0r** [comment]
WebKit2 does not have a `status-bar-text-changed` signal which is used for JS <-> Python communication. I don't see any other solution than to use window title for this purpose. The problem is that it is already used for an `evaluate_js` result.  Oh well, that means more hacks

**Comment by r0x0r** [comment]
BTW this is a C library that makes possible to parse JS result in Webkit2
https://github.com/saifulbkhan/wkjscore-result

If anyone finds a sane way to incorporate it into pywebview, maybe it can be used instead of window title tweaking

**Comment by mikaelho** [comment]
Could you use the WebKitNavigationPolicyDecision somehow, pass needed information in the URL, then reject the navigation request? (This is what we do with UIWebView in Pythonista.)

**Comment by r0x0r** [comment]
@mikaelho could you provide a url to example code?

**Comment by mikaelho** [comment]
I do not personally use GTK or WebKit2, but:

On Pythonista (Python on iOS) I use something like `window.location.href="js-callback:function_name?argument1=value1&argument2=value2";`. Then I get a `webview_should_start_load` callback on the Python side, with the url as an argument. I parse the function and arguments from the url, and return False to indicate that nothing should actually happen on the JS side.

Looking at the [WebKit2 Cookbook](https://wiki.gnome.org/Projects/WebKitGtk/ProgrammingGuide/Cookbook), under "Navigation, New Window and Response policies", it seems that the same approach would work, by [catching](https://lazka.github.io/pgi-docs/#WebKit2-4.0/classes/WebView.html#WebKit2.WebView.do_decide_policy) the `decide-policy` signal.

**Comment by shivaprsdv** [comment]
Looking at the original C documentation, `webkit_web_view_run_javascript()` has
a `callback` argument, which gets called after execution. Then there is a way to
get the result of the operation. Read:
> When the operation is finished, `callback` will be called. You can then call `webkit_web_view_run_javascript_finish()` to get the result of the operation.

Doesn't this exist in Pygobject @r0x0r? If it does, we could remove the hacks for both
`evaluate_js` and JS API.

**Comment by r0x0r** [comment]
I implemented retrieving JS value using the method described @mikaelho. Not sure how robust it is, but let's go with this one for the time being.

**Comment by r0x0r** [comment]
Another problem emerged with WebKit2. Namely web inspector throws an exception for local paths. Works fine for http://, but fails for file://
```
  File "/home/roman/pywebview/webview/gtk.py", line 107, in __init__
    self.webview.get_inspector().connect('inspect-web-view', self.on_inspect_webview)
TypeError: <WebKit2.WebInspector object at 0x7ffb992fec60 (WebKitWebInspector at 0x2e8e220)>: unknown signal name: inspect-web-view
```

**Comment by shivaprsdv** [comment]
WebKit2 Webview doesn't have the `inspect-web-view` signal. It shows the
inspector automatically, so we don't have to handle it manually.

I have removed `on_inspect_webview()` and related code (commiting after a
long time..🙂)

**Comment by r0x0r** [comment]
GTK Webkit2 upgrade is merged with master.

@shivaprsdv will you be able to give the WKWebView upgrade a go?

**Comment by shivaprsdv** [comment]
@r0x0r I have implemented a crude, basic transition. See `wkwebview` branch.
Lot of things still hanging, especially `evaluate_js()`. The WKWebView function
for evaluating JavaScript simply won't run; it keeps throwing some strange error
about the signature of the "block" it is supposed to accept as an argument. I've
dug to the deepest of PyObjC for a solution, to no avail. 😕

I am a bit busy with college ATM, progress might be slow.

**Comment by shivaprsdv** [comment]
OK I have (almost) ported `evaluate_js()`. I think the exception that pops up has
to do with an older version of PyObjC (I have been using the built in version, which
is 2.5.1 on High Sierra🤓). I just installed the latest (5.0b2) on Python3 and every-
thing works smooth. Need to test a recent build on Python2.x as well. Can you test
it on what you have got @r0x0r?

Meanwhile the `js_api` is still broken, because the old scripting interface does not
exist in WKWebView. Need to find an alternative.

**Comment by shivaprsdv** [comment]
Ported `js_api`, examples working fine. I have also tested both `evaluate_js()`
and `js_api` examples on Python 2.7 with PyObjC 5.0, running successfully.

BTW why is the Python 3 test commented out in `.travis.yml`? Pytest is broken
in my local machine ATM, so couldn't run tests.

**Comment by r0x0r** [comment]
@shivaprsdv Tests mostly fail. I did a quick investigation and `webView_didFinishNavigation_` is not fired for `load_html` calls in tests (nor HTML is loaded for that matter). `examples\load_html.py` works fine though. No idea why it happens.

There was an environmental problem with getting Python 3 up and running on Travis, so the test was disabled. Not sure if the problem still persists, but you may try re-enabling it.

**Comment by shivaprsdv** [comment]
Seems to be true. WKWebview itself uses a multi-processing architecture for
different tasks like networking, rendering, etc. But I don't know why this would
cause such an issue. May be it is a bug with WebKit itself? 😕

**Comment by shivaprsdv** [comment]
Another problem that is hindering the finishing of the upgrade are the methods
of `BrowserDelegate`. Most of them now has a `completionHandler` argument,
which is an ObjC block to be called once we finish the delegate action. No matter
if I call it or not, the webview freezes in the middle and sometimes Cocoa throws
an exception in the background that the `completionHandler` was not called. It
could well be an issue with the block-support of PyObjC. And some methods like
`printFrameView` are completely absent. 🤷‍♂️

**Comment by r0x0r** [comment]
I fixed the multiprocessing issue by changing the starting method to spawn. Some tests still fail though.

**Comment by r0x0r** [comment]
Bad news are that `multiprocessing.set_method_start` is available only from 3.4 onwards. Python 2 seems to have no way to specify how a new process is started. 
I have attempted to refactor tests to get rid of multiprocessing, but it has opened the whole new can of worms.

**Comment by r0x0r** [comment]
Managed to refactor tests, so that they work without spawning multiple processes now. All the tests pass now on macOS, both with Python 2 and 3. I haven't yet tested with other platforms. 

pyobjc requirement seems to be 3.1 or so. Hard-coding it in setup.py might be a good idea.

**Comment by r0x0r** [comment]
Also tests became super slow for macOS. No idea what causes it, but at least they pass.

**Comment by r0x0r** [comment]
A couple of findings.

* UIDelegate methods indeed do not get called. Changing UIDelegate's method signatures to something else results in an error, so at least methods are present and correct. This looks like a bug in objc or someplace else to me, as methods in other delegates work just fine

* `setPolicyDelegate` is removed in wkWebView and its methods now appear in NavigationDelegate. I haven't managed to get them working though. completionHandler must be called in the method body. Calling the handler as is results in a `TypeError: cannot call block without a signature`. I suspect wrapping a method inside `objc.selector` is the answer, but I have no clue how to do it yet.

**Comment by r0x0r** [comment]
Alert and confirm fixed now. 

Things to do:

- [ ] print functionality. Here is helper https://stackoverflow.com/questions/29236929/printing-contents-of-wkwebview-os-x
- [] _blank links. webView_decidePolicyForNavigationAction_navigationAction_decisionHandler_ seems to the right delegate
- [] right click menu is not disabled.

**Comment by shivaprsdv** [comment]
Unfortunately `alert` and `confirm` are not fixed by 0ffd11a. The webview UI freezes
after I dismiss the dialog.

STR: navigate to https://www.w3schools.com/jsref/tryit.asp?filename=tryjsref_alert in
the webview and click the alert button.

Result:
![test](https://user-images.githubusercontent.com/18247990/48667817-f9508380-eb04-11e8-8106-35a5e8b06f07.gif)

**Comment by r0x0r** [comment]
You are correct. My guess this is caused by the fact that completionHandler is not called. Still no idea how to call it properly, calling it as is results in a `{TypeError}cannot call block without a signature` exception.

**Comment by shivaprsdv** [comment]
Slight improvements in the situation: `evaluate_js` has been made to work with
older versions of PyObjC by manually loading the missing block signature. Should
be convenient for the end user, because macOS Mojave still ships with 2.5.1.

**Comment by shivaprsdv** [comment]
I have solved the `completionHandler` issue using a technique I found [here](https://github.com/munki/munki/commit/b9944e9#diff-f71080686b7bca85d6d6de30d45d5205R77). Now need to extend it to the NavigationDelagate methods as well.

**Comment by shivaprsdv** [comment]
Ported NavigationDelegate methods. `_blank` links are opened in browser
as before, while `window.open` does nothing, to be consistent with other
platforms.

**Comment by shivaprsdv** [comment]
Disabled context menu. And as for printing, WKWebView does not seem to
offer any solid solution as it predecessor. The solution offered in the above
SO does not seem to work very well.Everyone seems to be stuck on this.

**Comment by shivaprsdv** [comment]
I have implemented print functionality. Looks like I was wrong, `WKWebview`
_did_ have print functionality from [early on](https://github.com/WebKit/webkit/commit/0dfc67a174b79a8a401cf6f60c02150ba27334e5). But the catch is, it is still a private
API. It does the job best, but I don't know its acceptable to use private API.
Comments @r0x0r ?

**Comment by r0x0r** [comment]
I am fine with using the private API, as long as it gets the job done. By the way all the tests for macOS are failing with this error. Any idea?
```
E       Failed: 'NSUnknownKeyException - [<WKPreferences 0x7fe1723896c0> setValue:forUndefinedKey:]: this class is not key value coding-compliant for the key backspaceKeyNavigationEnabled.'
```

https://travis-ci.org/r0x0r/pywebview/jobs/464306221

**Comment by shivaprsdv** [comment]
Duh. It looks like this flag is available only from macOS Mojave onwards.😕
